pkg/apis/security/types.go

```diff
@@ -0,0 +1,307 @@
+package security
+
+import (
+	"crypto/sha1"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	cl "github.com/openshift/elasticsearch-proxy/pkg/handlers/clusterlogging/types"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+//DocType is the type of security document
+type DocType string
+
+const (
+	//DocTypeRoles are the security roles
+	DocTypeRoles DocType = "roles"
+
+	//DocTypeRolesmapping are the security mappings of users to roles
+	DocTypeRolesmapping DocType = "rolesmapping"
+)
+
+//ACLDocuments are the security documents
+type ACLDocuments map[DocType]ACLDocument
+
+//ACLDocument is a specific security document
+type ACLDocument interface {
+	//Iterate through entries that are expirable
+	Iterate() map[string]Expirable
+
+	//Remove a permission with the given name
+	Remove(name string)
+
+	//ToJson to covert the document to a JSON string
+	ToJson() (string, error)
+
+	//Type of the ACLDocument
+	Type() DocType
+
+	//Size provides the number of entries
+	Size() int
+
+	//Version of the document
+	Version() int
+}
+
+func (docs *ACLDocuments) Iterate() []ACLDocument {
+	entries := []ACLDocument{}
+	for _, doc := range *docs {
+		entries = append(entries, doc)
+	}
+	return entries
+}
+
+func (docs *ACLDocuments) Set(aclDoc ACLDocument) {
+	(*docs)[aclDoc.Type()] = aclDoc
+}
+
+func (docs *ACLDocuments) Roles() *Roles {
+	if val, exists := (*docs)[DocTypeRoles]; exists {
+		return val.(*Roles)
+	}
+	return nil
+}
+
+func (docs *ACLDocuments) RolesMapping() *RolesMapping {
+	if val, exists := (*docs)[DocTypeRolesmapping]; exists {
+		return val.(*RolesMapping)
+	}
+	return nil
+}
+
+//AddUser permissions to the ACL documents
+func (docs *ACLDocuments) AddUser(user *cl.UserInfo, expires int64) {
+	log.Tracef("Adding permissions for %s to expire at %v", user.Username, expires)
+	roleName := roleName(user)
+	role := Role{
+		ClusterPermissions: Permissions{"CLUSTER_MONITOR_KIBANA", "USER_CLUSTER_OPERATIONS"},
+		IndicesPermissions: newSecurityDocumentPermissions(user),
+	}
+	role.ExpiresInMillis = expires
+	docs.Roles().Set(roleName, role)
+	rolemapping := RoleMapping{
+		Users: []string{user.Username},
+	}
+	rolemapping.ExpiresInMillis = expires
+	docs.RolesMapping().Set(roleName, rolemapping)
+}
+
+//ExpirePermissions which are older then now
+func (docs *ACLDocuments) ExpirePermissions() {
+	log.Debug("Expiring permissions...")
+	now := time.Now().UnixNano() / int64(time.Millisecond)
```

Comment:
instead of doing this, can we instead use `time.Since`?

[1] https://golang.org/pkg/time/#Since

Reply:
updated to use 'Before' and "after"